dask/threaded.py

```diff
@@ -242,40 +241,41 @@ def execute_task(dsk, key, state, queue, results, lock):
     return
 
 
-def finish_task(dsk, key, result, state, results):
+def finish_task(dsk, key, result, state, results, lock):
     """
     Update executation state after a task finishes
 
     Mutates.  This should run atomically (with a lock).
     """
-    state['cache'][key] = result
-    if key in state['ready']:
-        state['ready'].remove(key)
-
-    for dep in state['dependents'][key]:
-        s = state['waiting'][dep]
-        s.remove(key)
-        if not s:
-            del state['waiting'][dep]
-            state['ready'].add(dep)
-
-    for dep in state['dependencies'][key]:
-        if dep in state['waiting_data']:
-            s = state['waiting_data'][dep]
+    with lock:
```

Comment:
One of the reasons I didn't put the lock into `finish_task` was that I wanted it to be easier to test.  Ideally we can isolate heavy state manipulation logic from the secondary logic of locking.  We should only merge the two if locking logic needs to interleave with state-manipulation logic (as is the case with fine-grained locking).

In this case it doesn't seem like `finish_task` needs to control the lock (nothing here is locked in a fine grained way).